### PR TITLE
[BACK-1746] Convert string dates to ISODATE in DB

### PR DESCRIPTION
**Pull request by Roukoswarf**
This has been tested using uploader and postman APITests along with manual testing, and covers all found cases using string dates.
Unit tests will most likely fail for now, as they manually specify date strings, and I believe it will be best to gather feedback on the approach, and discuss unit testing strategies before adjusting them.

### CODE REVIEW

**Review by toddkazakov — CHANGES_REQUESTED** [review]
`updateDatumInternal` needs to be updated so it uses dates instead of strings

**Review by toddkazakov — CHANGES_REQUESTED** [review]
`getDatumBefore` might also fail, because it doesn't filter by either `string` or `Date` time.

**Comment by Roukoswarf** [comment]
> `getDatumBefore` might also fail, because it doesn't filter by either `string` or `Date` time.

this should work, as its using the next record to find the previous, and that type would change dynamically, dates always sort after strings, so this should work even with mixed data.

File: lib/streamDAO.js [hunk 1]
```diff
@@ -53,6 +53,15 @@ function filterDatumForMongo(datum) {
       datum.deviceId = datum.deviceId.replace(BadDeviceIDPrefix, GoodDeviceIDPrefix);
       datum.id = schema.generateId(datum, schema.idFields(datum.type));
     }
+
+    // This is a migration step which converts times to native time objects before write to keep client changes minimal
+    // it would be better if this could be done right at input, but its simpler this way for now.
+    if ('time' in datum && typeof(datum.time) === 'string') {
```
Comment (toddkazakov) [inline lib/streamDAO.js:59]: This code is called when inserting or updating a datum. I believe deduplication will not work correctly if you first update a datum, then you try to reinsert the same one (with identical identity fields). Update doesn't change `id` and `_id` which are already generated based on the original string value of a date. When inserting a new one, you'll end up generating a new `id` and `_id` based on native date times, which will generate a different hash.

**Comment by toddkazakov** [comment]
There seems to be a problem when generating `basal.nextId`, which is also apparent from the failing unit tests. I don't exactly know what the problem is, without actually fixing it myself. The code is very difficult to follow.

**Comment by toddkazakov** [comment]
See `schema/schema.js#decepticon()`:
```js
function decepticon(datum, cb) {
    if (datum._userId == null) {
      return cb(new Error('_userId must be set on a datum'));
    }
    if (datum._groupId == null) {
      return cb(new Error('_groupId must be set on a datum'));
    }

    if (datum.time[datum.time.length - 1] !== 'Z') {
      datum.time = new Date(datum.time).toISOString();
    }

    try {
      transformer(datum, function(err, newDatum){
        if (err != null) {
          return cb(err);
        }
        var withId = attachId(newDatum);

        // Swap out cb so that we don't call it again if it throws out an error
        var theCallback = cb;
        cb = function(err) { if (err != null) { throw err; } };
        theCallback(null, withId);
      });
    } catch (e) {
      if (e.type === 'IllegalArgumentException') {
        e.statusCode = 400;
      }
      return cb(e);
    }
  }
```

This function wraps the transformer functions and in some cases might modify the time so it's set to a string value.

File: lib/streamDAO.js [hunk 2]
```diff
@@ -53,6 +53,25 @@ function filterDatumForMongo(datum) {
...
+
+	  if ('createdTime' in entry && typeof(entry.createdTime) === 'string') {
+	    entry.createdTime = new Date(entry.createdTime);
+	  }
+
+	  if ('archivedTime' in entry && typeof(entry.archivedTime) === 'string') {
+		entry.archivedTime = new Date(entry.archivedTime);
+	  }
+
+	  entry = entry.suppressed;
```
Comment (toddkazakov) [inline lib/streamDAO.js:73]: Should you convert dates in `datum.previous` the same you convert them in `datum.suspended`?

File: lib/streamDAO.js [hunk 3]
```diff
@@ -66,7 +85,7 @@ module.exports = function(mongoClient){
     }
 
     var clone = _.clone(datum);
```
Comment (toddkazakov) [inline lib/streamDAO.js:87]: Should you replace any time fields which are strings with native dates here?

**Review by toddkazakov — CHANGES_REQUESTED** [review]
Like we discussed few weeks ago on Slack, I believe you should convert the time keys back to strings right after they are fetched from the db to avoid changing any logic that relies on time attributes being strings (e.g. deduplication logic that includes the `time` attribute in the `_id` of the datum). I'm not sure this is a valid concern, but it feels like this is the right thing to do given that any issues with deduplication might go unnoticed.

Left some inline comments which also need to be addressed before this is ready for testing.

**Review by toddkazakov — APPROVED** [review]

**Pull request merged**